web/src/people.ts

```diff
@@ -1814,19 +1842,101 @@ export function is_displayable_conversation_participant(user_id: number): boolea
     return !is_valid_bot_user(user_id) && is_person_active(user_id);
 }
 
-export function initialize(my_user_id: number, params: StateData["people"]): void {
-    for (const person of params.realm_users) {
+export function populate_valid_user_ids(params: StateData["user_groups"]): void {
+    for (const user_group of params.realm_user_groups) {
+        if (user_group.is_system_group) {
+            valid_user_ids = valid_user_ids.union(new Set(user_group.members));
+        }
+    }
+}
+
+export function fetch_users_from_server(opts: FetchUserDataParams): void {
+    const params = {
+        user_ids: opts.user_ids,
+        client_gravatar: opts.client_gravatar,
+        include_custom_profile_fields: opts.include_custom_profile_fields,
+    };
+
+    channel.get({
+        url: "/json/users",
+        data: params,
+        success(data) {
+            if (opts.success) {
+                const members = user_fetch_response_schema.parse(data).members;
+                opts.success(members);
+            }
+        },
+        error(xhr: JQuery.jqXHR<unknown>) {
+            if (opts.error) {
+                opts.error(xhr);
+            }
+        },
+    });
+}
+
+export async function fetch_users(user_ids: Set<number>): Promise<UsersFetchResponse["members"]> {
+    const invalid_user_ids = user_ids.difference(get_valid_user_ids());
+    if (invalid_user_ids.size > 0) {
+        blueslip.info("Ignored invalid user_ids: " + [...invalid_user_ids].join(", "));
+    }
+    const user_ids_to_fetch = get_valid_user_ids().intersection(user_ids);
```

Comment:
Let's not call `get_valid_user_ids()` twice!

Reply:
Oops! Fixed.